dev/src/index.ts

```diff
@@ -1265,19 +1281,95 @@ export class Firestore implements firestore.Firestore {
 
     const writer = bulkWriter ?? this.getBulkWriter();
     writer._verifyNotClosed();
-    const docStream = this._getAllDescendants(
-      ref instanceof CollectionReference
-        ? (ref as CollectionReference<unknown>)
-        : (ref as DocumentReference<unknown>)
-    );
+
+    // A deferred promise that resolves when the recursive delete operation
+    // is completed.
     const resultDeferred = new Deferred<void>();
+
     let errorCount = 0;
     let lastError: GoogleError | BulkWriterError | undefined;
     const incrementErrorCount = (err: Error): void => {
       errorCount++;
       lastError = err;
     };
-    const onStreamEnd = (): void => {
+
+    // Whether all descendants for this query have been fetched.
+    let allDescendantsFetched = false;
+
+    // Whether a query stream is currently in progress. Only one stream can be
+    // run at a time.
+    let streamInProgress = false;
+
+    // The last document snapshot returned by the stream. Used to set the
+    // startAfter() field in the subsequent stream.
+    let lastDocumentSnap: QueryDocumentSnapshot;
+
+    // The number of pending BulkWriter operations. Used to determine when the
+    // next query can be run.
+    let pendingPromiseCount = 0;
+
+    // Deletes the provided reference, and starts the next stream if conditions
+    // are met.
+    const deleteRef = (docRef: DocumentReference): void => {
+      pendingPromiseCount++;
+      writer
+        .delete(docRef)
+        .then(() => pendingPromiseCount--)
+        .catch(err => {
+          pendingPromiseCount--;
+          incrementErrorCount(err);
+        })
+        .then(() => {
+          // We wait until the previous stream has ended in order to sure the
+          // startAfter document is correct. Starting the next stream while
+          // there are pending operations allows Firestore to maximize
+          // BulkWriter throughput.
+          if (
+            !allDescendantsFetched &&
+            !streamInProgress &&
+            pendingPromiseCount < RECURSIVE_DELETE_PENDING_OP_LIMIT
+          ) {
+            const nextDocStream = this._getAllDescendants(
+              ref instanceof CollectionReference
+                ? (ref as CollectionReference<unknown>)
+                : (ref as DocumentReference<unknown>),
+              /* startAfter= */ lastDocumentSnap
+            );
+            setupStream(nextDocStream);
+          }
+        });
```

Comment:
This might be easier to reason about if you move the logic out of the `then` block of the delete and instead invoked a separate function. This is about the query itself and not the individual delete anymore.

I also think that this logic is not complex enough that it likely warrants its own class, which would allow you to use proper methods.

Reply:
Moved the logic into a separate class.